File: book/petrol-pump/chapter3/section4/ShowDollarsPump.java
package chapter3.section4;

import pump.*;
import chapter3.section2.LifeCycle;
import sodium.*;
import java.util.Optional;

public class ShowDollarsPump implements Pump {
    public Outputs create(Inputs inputs) {
        LifeCycle lc = new LifeCycle(inputs.eNozzle1,
                                     inputs.eNozzle2,
                                     inputs.eNozzle3);
        Fill fi = new Fill(lc.eStart.map(u -> Unit.UNIT),
                           inputs.eFuelPulses, inputs.calibration,
                           inputs.price1, inputs.price2, inputs.price3,
                           lc.eStart);
        return new Outputs()
            .setDelivery(lc.fillActive.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleCostLCD(fi.dollarsDelivered.map(
                q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(priceLCD(lc.fillActive, fi.price, Fuel.ONE,
                inputs))
            .setPriceLCD2(priceLCD(lc.fillActive, fi.price, Fuel.TWO,
                inputs))
            .setPriceLCD3(priceLCD(lc.fillActive, fi.price, Fuel.THREE,
                inputs));
    }

    public static Behavior<String> priceLCD(
            Behavior<Optional<Fuel>> fillActive,
            Behavior<Double> fillPrice,
            Fuel fuel,
            Inputs inputs) {
        Behavior<Double> idlePrice;
        switch (fuel) {
            case ONE:   idlePrice = inputs.price1; break;
            case TWO:   idlePrice = inputs.price2; break;
            case THREE: idlePrice = inputs.price3; break;
            default:    idlePrice = null;
        }
        return Behavior.lift((oFuelSelected, fillPrice_, idlePrice_) ->
            oFuelSelected.isPresent()
                ? oFuelSelected.get() == fuel
                                      ? Formatters.formatPrice(fillPrice_)
                                      : ""
                : Formatters.formatPrice(idlePrice_),
            fillActive,
            fillPrice,
            idlePrice);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section6/ShowDollarsPump.cs
using System;

using PetrolPump.Chapter3.Section3;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section6
{
  public class ShowDollarsPump : IPump 
  {
    public Outputs Create(Inputs inputs) {
        LifeCycle lc = new LifeCycle(inputs.ENozzle1,
                                     inputs.ENozzle2,
                                     inputs.ENozzle3);
        Fill fi = new Fill(lc.EStart.Map(u => Unit.UNIT),
                           inputs.EFuelPulses, inputs.Calibration,
                           inputs.Price1, inputs.Price2, inputs.Price3,
                           lc.EStart);
        return new Outputs()
            .SetDelivery(lc.FillActive.Map(
                of =>
                    of.Equals(Optional<Fuel>.Of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.Equals(Optional<Fuel>.Of(Fuel.TWO)) ? Delivery.FAST2 :
                    of.Equals(Optional<Fuel>.Of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .SetSaleCostLcd(fi.DollarsDelivered.Map(
                q => Formatters.FormatSaleCost(q)))
            .SetSaleQuantityLcd(fi.LitersDelivered.Map(
                q => Formatters.FormatSaleQuantity(q)))
            .SetPriceLcd1(PriceLCD(lc.FillActive, fi.Price, Fuel.ONE,
                inputs))
            .SetPriceLcd2(PriceLCD(lc.FillActive, fi.Price, Fuel.TWO,
                inputs))
            .SetPriceLcd3(PriceLCD(lc.FillActive, fi.Price, Fuel.THREE,
                inputs));
    }

    public static Behavior<String> PriceLCD(
            Behavior<Optional<Fuel>> fillActive,
            Behavior<Double> fillPrice,
            Fuel fuel,
            Inputs inputs) {
        Behavior<Double> idlePrice;
        switch (fuel) {
            case Fuel.ONE:   idlePrice = inputs.Price1; break;
            case Fuel.TWO:   idlePrice = inputs.Price2; break;
            case Fuel.THREE: idlePrice = inputs.Price3; break;
            default:    idlePrice = null; break;
        }
        return Behavior<string>.Lift((oFuelSelected, fillPrice_, idlePrice_) =>
            oFuelSelected.IsPresent
                ? oFuelSelected.Get() == fuel
                                      ? Formatters.FormatPrice(fillPrice_)
                                      : ""
                : Formatters.FormatPrice(idlePrice_),
            fillActive,
            fillPrice,
            idlePrice);
    }
}

}
